File: cpp/src/signal_module.cpp
//
// Created by Kanishka Roy on 7/23/26.
//
// C++ port of python/signal_module.py

#include "../include/phy/signal_module.h"

#include <complex>
#include <set>
#include <stdexcept>

#include "phy/fec_module.h"
#include "phy/interleaver_module.h"
#include "phy/modulation_module.h"
#include "phy/ofdm_module.h"
#include "phy/pilot_utils.h"

namespace wifi80211a
{
    namespace
    {
        // SIGNAL is always one BPSK symbol: NCBPS=48, NBPSC=1.
        constexpr int kSignalNCBPS = 48;
        constexpr int kSignalNBPSC = 1;

        int even_parity_(const std::vector<int>& bits, int count)
        {
            int parity = 0;
            for (int i = 0; i < count; i++)
            {
                parity ^= bits[i];
            }
            return parity;
        }
    }

    complexVector create_signal_header(const LinkSettings& linkSettings, int& psduLengthOctets)
    {
        std::vector<int> signalBits;

        const int rateValue = inverseRateMap.at({linkSettings.getModulationType(), linkSettings.getCodingRate()});
        const std::vector<int> rateBits = int_to_bits(static_cast<uint64_t>(rateCodeword.at(rateValue)), 4);
        signalBits.insert(signalBits.end(), rateBits.begin(), rateBits.end());

        signalBits.push_back(0); // reserved

        const std::vector<int> lengthBits = int_to_bits(static_cast<uint64_t>(psduLengthOctets), 12);
        signalBits.insert(signalBits.end(), lengthBits.begin(), lengthBits.end());

        signalBits.push_back(even_parity_(signalBits, 17)); // even parity over bits 0-16

        // append_convolutional_tail_ supplies the 6 zero SIGNAL TAIL bits (18-23),
        const std::vector<int> tailedBits = append_convolutional_tail_(signalBits);
        const std::vector<int> encodedBits = convolutional_encoder_mother_(tailedBits);

        const std::vector<int> interleavedBits = interleave(encodedBits, kSignalNCBPS, kSignalNBPSC);

        const complexVector modulatedSymbols = map_bits_to_constellation(
            interleavedBits, ModulationTypes::BPSK, kSignalNBPSC);

        // OFDM modulate (pilots + IFFT + cyclic prefix) -> one 80-sample symbol
        PilotLFSR pilotLfsr;
        return modulate(linkSettings, modulatedSymbols, pilotLfsr);
    }

    SignalOutput decode_signal_header(LinkSettings& linkSettings, complexVector& signalHeader, complexVector& H)
    {
        const OFDMDemodResult demod = demodulate(linkSettings, signalHeader);

        const int nFFT = linkSettings.getNFFT();
        const std::vector<int> pilotPositions = linkSettings.getPilotPositions();
        const std::set<int> pilotSet(pilotPositions.begin(), pilotPositions.end());

        // Data-subcarrier bins in the same order create_signal_header assigned them
        complexVector dataBins;
        dataBins.reserve(kSignalNCBPS);
        for (int k = -26; k <= 26; k++)
        {
            if (k == 0 || pilotSet.contains(k)) continue;

            const int fftBin = fft_bin_from_subcarrier(k, nFFT);
            std::complex<double> sample = demod.freq_bins[fftBin];
            if (!H.empty())
            {
                const std::complex<double> H_k =
                    (std::abs(H[fftBin]) > 1e-10) ? H[fftBin] : std::complex<double>(1.0, 0.0);
                sample /= H_k;
            }
            dataBins.push_back(sample);
        }

        const std::vector<int> codedBits = map_constellation_to_bits(dataBins, ModulationTypes::BPSK, kSignalNBPSC);
        const std::vector<int> deinterleavedBits = deinterleave(codedBits, kSignalNCBPS, kSignalNBPSC);

        const std::vector<int> maskBits(deinterleavedBits.size(), 1);
        std::vector<int> decodedBits = viterbi_decode_(deinterleavedBits, maskBits);
        decodedBits.resize(18); // drop the 6 SIGNAL TAIL bits

        const int rateCodewordBits = decodedBits[0] | (decodedBits[1] << 1) | (decodedBits[2] << 2) | (decodedBits[3] << 3);
        const int parityBit = decodedBits[17];

        const int expectedParity = even_parity_(decodedBits, 17);
        if (expectedParity != parityBit)
        {
            throw std::runtime_error("SIGNAL field parity check failed");
        }

        const auto [modulationType, codingRate] = rateMap.at(rateCodewordToValue.at(rateCodewordBits));

        int psduLengthOctets = 0;
        for (int i = 0; i < 12; i++)
        {
            psduLengthOctets |= decodedBits[5 + i] << i;
        }

        linkSettings.changeModulationType(modulationType);
        linkSettings.changeCodingRate(codingRate);

        return SignalOutput{modulationType, codingRate, psduLengthOctets};
    }
}

File: python/signal_module.py
import numpy as np

from config import *
from fec_module import append_convolutional_tail, convolutional_encoder_mother, viterbi_decode
from interleaver_module import deinterleave, interleave
from modulation_module import map_bits_to_constellation, map_constellation_to_bits
from ofdm_module import OFDMModule
from utils import fft_bin_from_subcarrier, int_to_bits


def create_signal_header(
    linkSettings: LinkSettings, psduLengthOctets: int
) -> np.ndarray:
    """Builds and OFDM-modulates the 24-bit SIGNAL field (Sec. 17.3.4).

    SIGNAL is always BPSK, R=1/2, and is never scrambled or punctured,
    regardless of the RATE used for the DATA field that follows it.

    RATE is taken from linkSettings.modulationType / linkSettings.codingRate
    (same as C++ create_signal_header).
    """
    signalBits = []
    rateValue = rateMapInverse[(linkSettings.modulationType, linkSettings.codingRate)]
    signalBits.extend(int_to_bits(rateCodeword[rateValue], 4))
    signalBits.append(0)  # reserved
    signalBits.extend(int_to_bits(psduLengthOctets, 12))
    signalBits.append(sum(signalBits[:17]) % 2)  # even parity over bits 0-16
    signalBits = np.array(signalBits, dtype=np.int8)

    # append_convolutional_tail supplies the 6 zero SIGNAL TAIL bits (18-23),
    # completing the 24-bit field before the mother-rate 1/2 conv. encode.
    tailedBits = append_convolutional_tail(signalBits)
    encodedBits = np.array(convolutional_encoder_mother(tailedBits))

    # Interleave (SIGNAL is always one BPSK symbol: NCBPS=48, NBPSC=1)
    interleavedBits = interleave(encodedBits, n_cbps=48, n_bpsc=1)

    modulatedSymbols = np.array(
        map_bits_to_constellation(interleavedBits, ModulationTypes.BPSK, 1),
        dtype=np.complex64)

    # OFDM modulate (pilots + IFFT + cyclic prefix) -> one 80-sample symbol
    ofdm = OFDMModule(linkSettings)
    return ofdm.modulate(modulatedSymbols)


def decode_signal_header(
    linkSettings: LinkSettings, signalHeader: np.ndarray, H: np.ndarray | None = None
) -> tuple[str, str, int]:
    """Inverse of create_signal_header: OFDM demod -> (optional LTF equalize)
    -> BPSK demap -> deinterleave -> Viterbi decode (R=1/2, nothing punctured)
    -> RATE/LENGTH/parity.

    On success, updates linkSettings modulation/coding to the decoded RATE
    (mirrors C++ decode_signal_header).

    H — optional LTF-derived channel estimate (see sync_module); the SIGNAL
        symbol rides through the same channel as the DATA field, so it must
        be equalized the same way before demapping.

    Returns (modulationType, codingRate, psduLengthOctets) for the DATA field
    that follows this SIGNAL symbol.
    """
    ofdm = OFDMModule(linkSettings)
    demod = ofdm.demodulate(signalHeader)

    # Data-subcarrier bins in the same order create_signal_header assigned them
    dataBins = []
    for k in range(-26, 27):
        if k == 0 or k in linkSettings.pilotPositions:
            continue
        fftBin = fft_bin_from_subcarrier(k, linkSettings.nFFT)
        sample = demod.freqBins[fftBin]
        if H is not None:
            H_k = H[fftBin] if np.abs(H[fftBin]) > 1e-10 else 1.0
            sample = sample / H_k
        dataBins.append(sample)

    codedBits = map_constellation_to_bits(dataBins, ModulationTypes.BPSK, 1)
    deinterleavedBits = deinterleave(np.array(codedBits), n_cbps=48, n_bpsc=1)

    maskBits = np.ones(len(deinterleavedBits), dtype=int)
    decodedBits = viterbi_decode(deinterleavedBits, maskBits)[:18]  # drop 6 tail bits

    rateBits = decodedBits[0:4]
    lengthBits = decodedBits[5:17]
    parityBit = decodedBits[17]

    rateCodewordBits = sum(int(b) << i for i, b in enumerate(rateBits))
    if rateCodewordBits not in rateCodewordToValue:
        raise ValueError(
            f"Invalid SIGNAL RATE codeword 0b{rateCodewordBits:04b} "
            f"({rateCodewordBits}); expected one of "
            f"{sorted(f'0b{c:04b}' for c in rateCodewordToValue)}"
        )
    modulationType, codingRate = rateMap[rateCodewordToValue[rateCodewordBits]]
    psduLengthOctets = sum(int(b) << i for i, b in enumerate(lengthBits))

    expectedParity = sum(decodedBits[:17]) % 2
    if expectedParity != parityBit:
        raise ValueError("SIGNAL field parity check failed")

    linkSettings.change_modulation_type(modulationType)
    linkSettings.change_coding_rate(codingRate)

    return modulationType, codingRate, psduLengthOctets
